ENH: supply our version of numpy.pxd, requires cython>=0.29

mattip opened the pull request:
Fixes #11803

Adds our own `numpy.pxd`, copied from the cython `__init__.pxd`. It will only be used if the numpy include path from `numpy.get_includes()` is added to the cython compile flags, otherwise the cython one will still be used.

Modifications were needed to `mtrand.pyx` since `ndarray.shape` is a C `npy_intp` in the newer `numpy.pxd`, not a python-level sequence attribute.

mattip committed:
ENH: supply our version of numpy.pxd, adjust as needed

mattip commented:
Also the `numpy.pxd` uses the Cython `check_size ignore` syntax to avoid the "size changed" warnings, which requires cython 0.29 and above. Once downstream users update, we can remove the warning filters from `numpy.__init__.py`. xref #11788

eric-wieser commented on numpy/core/include/numpy/numpy.pxd:
```diff
@@ -0,0 +1,1053 @@
+        NPY_SEARCHRIGHT
+
+    enum:
+        # DEPRECATED since NumPy 1.7 ! Do not use in new code!
```
Given we just added this, is there any reason to keep the deprecated headers?

mattip replied:
As long as we have `npy_1_7_deprecated_api.h` for people who are still using the old API (scipy, pandas, ...), we should reflect old values in Cython.

mattip replied:
In order to truly deprecate the old API, we need either cython/cython#2640 or to revert the attribute hiding PyArray_FUNCTIONS since cython does not convert python attribute lookup `array.ndim` into a getter function `PyArray_NDIMS(array)`

eric-wieser replied:
Why do we need `array.ndim` to be converted into `PyArray_NDIMS(array)`? It would be good to have written down somewhere what problem is trying to be solved here.

mattip replied:
[cython class overview](https://cython.readthedocs.io/en/latest/src/tutorial/cdef_classes.html) gives the motivation: write python-like code, run it through the cython compiler and get efficient C code out. [Extension type spec](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#external-extension-types) dives into the details of how to map python object attributes to C struct fields.

We try to hide the C struct fields in `PyArrayObject_fields` behind C getter functions to the opaque `PyArrayObject` struct, so the techniques described in the links above are not sufficient.

mattip commented:
xref cython/cython#2498

mattip commented:
after cherry picking the changes from #12299, travis and lgtm are now passing

charris commented on .travis.yml:
```diff
@@ -41,7 +41,7 @@ matrix:
       env: INSTALL_PICKLE5=1
     - python: 3.6
-      env: USE_CHROOT=1 ARCH=i386 DIST=bionic PYTHON=3.6
+      env: USE_CHROOT=1 ARCH=i386 DIST=bionic PYTHON=python3 PIP=pip3
```
Don't need this, it is already done in `tools/travis-test.sh`. Please, one thing per PR.

mattip replied:
rebased to fix this

eric-wieser commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -35,6 +33,17 @@ cdef extern from "math.h":
 
+# Put initarray first since it includes the NPY_NO_DEPRECATED_API definition
+cdef extern from "Python.h":
+    ctypedef int Py_intptr_t
```
Is the fact that this is missing from the `Python.pxi` an upstream bug?

mattip replied:
which `Python.pxi`? The `cdef extern from "Python.h"` tells Cython to add `#include Python.h` at this point in the output C file, and promises that that header file (or one of the headers included by it) will include a `typedef int Py_intptr_t` statement.

eric-wieser replied:
[this one](https://github.com/numpy/numpy/pull/12284/files#diff-70d0fdd5755b6d140e4f09d9de3da144R24) - my assumption was that this header already includes all of the `cdef externs` that match to `Python.h`. Is that not true?

eric-wieser commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -384,7 +389,7 @@ cdef object discnp_array(rk_state *state, rk_discnp func, object size,
         array = <ndarray>np.empty(size, dtype=int)
         multi = <broadcast>np.broadcast(on, op, array)
-        if multi.shape != array.shape:
+        if multi.shape != getattr(array, 'shape'):
```
Do `x.attr` and `getattr(x, 'attr')` behave differently in cython? If so, that's pretty scary.

mattip replied:
`array.shape` becomes [`array.dimensions`](https://github.com/numpy/numpy/pull/12284/files#diff-35cf7adc200725d2d15266c5e58a07f3R253) in C, but `multi.shape` is `PyObject_GetAttrString(multi, "shape")`, so we either need to make sure cython uses `PyObject_GetAttrString(array, "shape")` and not the fast C struct field reference.

eric-wieser replied:
If we ask for the non-deprecated API in this file, doesn't that resolve the problem by making Cython unaware of the underlying struct members?

mattip replied:
Using the non-deprecated API would require converting `array.shape` to `npy_intp* PyArray_SHAPE(array)`. cython/cython#2640 aims to automatically do the conversion. This would still not return a `PyObject*`.

mattip replied:
That is an existing issue/feature of cython, not part of any changes made for this PR. For instance, it is worked around in scipy [here](https://github.com/scipy/scipy/blob/v1.1.0/scipy/ndimage/src/_ni_label.pyx#L210) which shows a more cythonic construct than the one I used but produces the same C code. I will change this PR to use `(<object>array).shape` casting.

eric-wieser commented on numpy/core/include/numpy/numpy.pxd:
```diff
@@ -0,0 +1,1053 @@
+    else:
+        return ()
+
+cdef inline char* _util_dtypestring(dtype descr, char* f, char* end, int* offset) except NULL:
```
this function worries me - can we not reuse our PEP3118 format string builder, rather than having a second implementation?

mattip replied:
I switched this all for `PyObject_GetBuffer()` which calls `array_getbuffer`, and a NOOP `__release_buffer__`, with a note explaining why it is a NOOP currently.

At some point if we drop our `_buffer_info_cache` we should revisit this code

jdemeyer commented:
I don't understand why you hide `numpy.pxd` in the `include` subdirectory.

Typically, the layout of `.pxd` files follows that of the Python project. Think of a `cimport` as analogous to an `import`. So, if you expect users to do `cimport numpy`, then it makes more sense to move `numpy.pxd` to a top-level `__init__.pxd` file (more or less the way it is in Cython).

mattip commented:
@jdemeyer whatever works correctly. I am trying to get this to work with pandas, so that `python setup.py build_ext --inplace --force` will find the numpy `pxd` before the one in `Cython/Include/numpy/__init__.pxd`. It seems no matter what I do the cython one is found first (I added a line like `x = xxxxxx` in the file to make compilation fail if the wrong one is chosen). What I tried:
- using the `include_path = pkg_resources.resource_filename('numpy', 'core/include')`
- putting it in `site-packages/numpy/__init__.pxd` (your suggestion)

Neither of those helped. But just now I tried using `include_path = pkg_resources.resource_filename('numpy', 'core/include/numpy')` which seems to work

rgommers commented:
> It seems no matter what I do the cython one is found first 

Shouldn't this be considered a Cython bug? `numpy/__init__.pxd` should be found before Cython's `numpy.pxd` I'd think. Maybe there are other packages where Cython needs to override something, but if that behavior is necessary for backwards compat or something then it would be nice if we had a way to turn it off.

> But just now I tried using `include_path = pkg_resources.resource_filename('numpy', 'core/include/numpy')` which seems to work

That really should not be necessary; that way whatever we ship is not very useful, since most people will never get it. Plus, I'd really like to avoid having to recommend `pkg_resources` .....

mattip commented:
I moved the file to `numpy/__init__.pxd`, and added an explanation to the release notes. Maybe cython could fix the problem for 3.0?

mattip commented:
xref cython/cython#2905

mattip commented:
This only passes tests because we do not have `cimport numpy` in our codebase

h-vetinari commented:
Any news on this? How realistic is it that this makes it into 1.17? :)

mattip commented:
Even if we merged this, cython still will not use it until a release that includes cython/cython#2910, which is only marked for cython 3.0. I think we should hold off until we can test this with that fix. Otherwise we risk breaking a something in an uncontrolled fashion.

scoder commented:
What is the timeframe for the 1.17 release? If Cython pushes a 0.29.11 release with the include path fix now, would it allow this change to be merged?

scoder commented:
It seems to me that, strictly speaking, it's not really a requirement to release the include path change in Cython before NumPy 1.17. It can just ship its own declarations now, and as soon as Cython releases the fix, they will be picked up. As long as the declarations don't change, that shouldn't make a difference to users.

h-vetinari commented:
For reference cython 0.29.11 has been released two weeks ago, and includes cython/cython#2993.

Enough to try a `cimport numpy` and continue this PR? :)

h-vetinari commented:
@mattip 
Seems 1.17 is pretty close to release. Would you mind if I rebased this in a new PR (if it's still realistic at all to make it before the cutoff)?

mattip commented:
rebased and squashed, but I am not optimistic about 1.17 which is already has released rc2

mattip commented:
I tested that our version is used by adding invalid code into `<site-packages>/Cython/Include/numpy/__init__.pxd`, for instance `xxxxxxxx = xxxxxxxxxxx` at the top of the file

h-vetinari commented:
> @mattip: rebased and squashed, but I am not optimistic about 1.17 which is already has released rc2

I understand it's late in the cycle, but if nothing breaks (see test suite), this should have 0 user-facing impact. Maybe have an rc3 to be sure, but otherwise, this should be possible, I hope. :)

mattip commented:
Travis is not feeling well today, python 3.5 tests are failing to download a proper python among other infra failures

h-vetinari commented on tools/cythonize.py:
```diff
@@ -71,13 +73,9 @@ def process_pyx(fromfile, tofile):
-            required_version = LooseVersion('0.28')
+        # Require at least 0.29 for the ctypdef check_size option
+        # cython/cython#2627
+        required_version = LooseVersion('0.29')
```
Shouldn't this be 0.29.11? @mattip

mattip replied:
yes, fixed

mattip commented:
removed the WIP tag

seberg reviewed (approved):
It seems right to me to move it into numpy (and I think there is general consensus here). Since it moved around, just to be sure: `numpy.get_includes()` does not need to be modified for dependencies to pick up our `numpy.pxd` file?

Unfortunately, there are some merge conflicts (I can fix them up maybe though).

mattip commented:
- rebased off master
- fixed release note: no changes in include path are needed, the pxd file is picked up via `cimport numpy` after cython/cython#

mattip commented:
- rebased off master
- fixed release note: no changes in include path are needed, the pxd file is picked up via `cimport numpy` after cython/cython#2910 which has been backported to the current release

seberg commented on doc/release/upcoming_changes/12284.new_feature.rst:
```diff
@@ -0,0 +1,6 @@
+
+Add our own ``*.pxd`` cython import file
+--------------------------------------------
+Added a ``numpy/__init__.pxd`` file. It will be used for `cpython import numpy`
```
```suggestion
Added a ``numpy/__init__.pxd`` file. It will be used for `cimport numpy as npc`
```
is what you mean I think? And I am not sure we still need the search path link (at least not as a link).

mattip replied:
fixed, thanks

seberg commented:
Is there a difference between this `pyd` and the one in cython that can explain the windows crash? Or is that a random crash due to OpenBLAS or so?

mattip commented:
seems like a random crash

seberg commented:
Thanks, lets put this in then.

jakirkham commented:
Thanks all for the work on this! 😀

scoder commented:
@mattip I checked in my local numpy 1.18.3 installation and it lacks the `numpy/__init__.pxd` file that this PR added. My guess is that it still needs to be added to the distutils `package_data` in order to be included not just in the sdist but also in binary distributions.

scoder commented:
Although … the `.pxd` files in the `numpy/random/` directory _are_ included. So, why it `numpy/__init__.pxd` not in my wheel?

jakirkham commented:
Not sure. Looks like [they should all be there]( https://github.com/numpy/numpy/blob/v1.18.4/MANIFEST.in#L17-L19 ).

scoder commented:
> they should all be there

Yes, that's the MANIFEST file that's used for the sdist. It doesn't have an effect on the binary distributions. However, I couldn't find any package data configuration or so in NumPy's `setup.py`. So how do the files in `numpy.random` get included?

scoder commented:
Ah, [found it](https://github.com/numpy/numpy/blob/master/numpy/random/setup.py), I think.

mattip commented:
@scoder gh-16162, thanks for the heads up.

scoder commented on numpy/__init__.pxd:
```diff
@@ -0,0 +1,978 @@
+            dtype descr  # deprecated since NumPy 1.7 !
+            PyObject* base
+
+        # Note: This syntax (function definition in pxd files) is an
+        # experimental exception made for __getbuffer__ and __releasebuffer__
+        # -- the details of this may change.
+        def __getbuffer__(ndarray self, Py_buffer* info, int flags):
+            PyObject_GetBuffer(<object>self, info, flags);
+
+        def __releasebuffer__(ndarray self, Py_buffer* info):
+            # We should call a possible tp_bufferrelease(self, info) but no
+            # interface to that is exposed by cython or python. And currently
+            # the function is NULL in numpy, we rely on refcounting to release
+            # info when self is collected
+            pass
```
NumPy can just leave these out entirely. They were added in the time when Python 2.6/3.0 were still new and the average NumPy installation on user side didn't support the PEP 3118 buffer protocol yet. The times have changed a bit since then. Even Cython can probably drop these by now. :)

mattip replied:
Cython first. I think there are tests to verify that the buffer protocol works for ndarray in cython.

scoder replied:
The buffer protocol _does_ work for ndarrays, because NumPy supports it natively. The implementation through `PyObject_GetBuffer()` shows this clearly (Cython still uses a real one). In fact, since `__releasebuffer__` is empty, releasing buffers will not even work here.

scoder replied:
I removed both methods in https://github.com/cython/cython/commit/b7949976596d2a08e98990ac88b034410c62e39b

mattip replied:
removing here too

jbrockmendel commented on numpy/__init__.pxd:
```diff
@@ -0,0 +1,978 @@
+    object PyArray_Choose (ndarray, object, ndarray, NPY_CLIPMODE)
+    int PyArray_Sort (ndarray, int, NPY_SORTKIND)
+    object PyArray_ArgSort (ndarray, int, NPY_SORTKIND)
+    object PyArray_SearchSorted (ndarray, object, NPY_SEARCHSIDE)
```
@seberg I think this is inaccurate xref https://stackoverflow.com/questions/28184211/calling-pyarray-searchsorted-from-cython-3-or-4-arguments

charris replied:
Hmm, looks like you are right. Since 2012 it has been
```
PyArray_SearchSorted(ndarray, object,  NPY_SEARCHSIDE, object)
```

seberg replied:
Was it safe to add an argument when its NULL?! Or was this just always wrong?
